Sources/Hyperspace/Service/Certificate Pinning/PinningConfiguration.swift

```diff
@@ -0,0 +1,141 @@
+        public init(domain: String, enforced: Bool = true, includeSubdomains: Bool = true, encodedPinningHashes: [String], expiration: Date? = nil) {
+            let pinningHashes = encodedPinningHashes.compactMap { Data(base64Encoded: $0) }
+            self.init(domain: domain, enforced: enforced, includeSubdomains: includeSubdomains, pinningHashes: pinningHashes, expiration: expiration)
+        }
+        
+        /// Creates an instance of `DomainConfiguration` which will govern how any presented SSL certificate from that domain will be pinned.
+        ///
+        /// - Parameters:
+        ///   - domain: The domain for which this configuration is responsible.
+        ///   - enforced: When this value is set to true, a failure to pin a request will cause the connection to be blocked. Defaults to true.
+        ///   - includeSubdomains: Include any subdomains of the given domain in the pinning process. Defaults to true.
+        ///   - pinningHashes: An array of public key hashes that will be used to verify the presented SSL certificate.
+        ///   - expiration: The expiration date of the SSL Certificate to be pinned. After this date, pinning will not be attempted.
+        init(domain: String, enforced: Bool = true, includeSubdomains: Bool = true, pinningHashes: [Data], expiration: Date? = nil) {
+            self.domain = domain
+            self.enforced = enforced
+            self.includeSubdomains = includeSubdomains
+            self.pinningHashes = pinningHashes
+            self.expiration = expiration
+        }
+        
+        // MARK: Interface
+        func shouldValidateCertificate(forHost host: String) -> Bool {
+            guard includeSubdomains else { return domain == host }
+            return host.contains(domain)
+        }
+        
+        func shouldValidateCertificate(forHost host: String, at date: Date) -> Bool {
+            guard shouldValidateCertificate(forHost: host), let expired = expiration else { return true }
+            return date < expired
+        }
+        
+        func validate(against remoteCertificate: SecCertificate) -> Bool {
+            let pinningHash = try? CertificateHasher.pinningHash(for: remoteCertificate)
+            return pinningHash.map(pinningHashes.contains) ?? false
+        }
+        
+        var dispositionForFailedValidation: URLSession.AuthChallengeDisposition {
+            if enforced {
+                return .cancelAuthenticationChallenge
+            }
+            
+            debugPrint("SSL certificate pin failed for host: '\(domain)'. Configuration set to allow connection anyway.")
+            return .performDefaultHandling
+        }
+        
+        // MARK: Hashable
+        public var hashValue: Int {
+            return domain.hashValue
+        }
+        
+        public static func == (lhs: DomainConfiguration, rhs: DomainConfiguration) -> Bool {
+            return lhs.domain == rhs.domain
```

Comment:
Currently we do an Array -> Set conversion when you initialize a configuration so we don't end up with multiple configs for a single domain. The default equatable conformance doesn't catch them as equal unless we're only checking domain. I can probably do a manual unique check instead.

Reply:
Ah, so it's the `domainConfigurations` property that doesn't provide a good default `Equatable` implementation? That's unfortunate since it seems like the whole idea behind Sets is to make that stuff easier.

Reply:
Well, I guess usually it'd be fine - it was just with the default conformance if I had two different `DomainConfiguration` objects with the same `domain` property, they wouldn't be seen as equal and so the Set wouldn't do the filtering for me. 

It feels like enough of a 'why would you do this?' that I've gotten rid of the change and put an assertion and a comment in the headers instead.